FRONTEND AND BACKEND MERGE/IMPLEMENTATION OF SAGA

TarikVelic opened the pull request:
--------------------------------------------------------------------
 MAIN
 *  CONNECTED FRONTEND AND BACKEND / PART OF LOGIN IMPLEMENTATION OF REDUX - SAGA

--------------------------------------------------------------------
 OTHER
 *  //// ////

--------------------------------------------------------------------

TarikVelic committed:
FRONTEND AND BACKEND MERGE/IMPLEMENTATION OF SAGA

--------------------------------------------------------------------
 MAIN
 *  CONNECTED FRONTEND AND BACKEND / PART OF LOGIN
    IMPLEMENTATION OF REDUX - SAGA

--------------------------------------------------------------------
 OTHER
 *  ////
    ////

--------------------------------------------------------------------

becarselman commented on src/actions/authActions.js:
```diff
@@ -0,0 +1,54 @@
+import { AUTHENTICATE_USER, GET_ERRORS, LOGOUT_USER, LOGIN_ERROR, LOGIN_REQUEST } from "./types";
+import authService from "../service/authService";
+
+export const authenticateUser = (email, password) => dispatch => {
```
Bussines logic need to be done in sagas

becarselman commented on src/actions/authActions.js:
```diff
@@ -0,0 +1,54 @@
+    .then(response => {
+      const { token, user } = response;
+      localStorage.setItem("jwtToken", token);
+      dispatch({  
```
actions need to be function that return type and payload. You don't dispatch anything here

becarselman commented on src/actions/authActions.js:
```diff
@@ -0,0 +1,54 @@
+};
+
+export const logoutUser = () => dispatch => {
+  localStorage.removeItem("jwtToken");
```
same as previous com.

becarselman commented on src/actions/types.js:
```diff
@@ -0,0 +1,7 @@
+export const AUTHENTICATE_USER = 'AUTHENTICATE_USER';
```
You can create one object ```
export type ={
AUTHENTICATE_USER : 'AUTHENTICATE_USER';
}
```

and then use like type.AUTHENTICATE_USER

becarselman commented on src/components/loginform/LoginForm.js:
```diff
@@ -1,7 +1,37 @@
+  const password = formData.get('login_password');
+  authService.login(email, password)
+    .then((response) => {
+      window.location.href = '/welcome';
```
change this. use something else for navigation

becarselman commented on src/components/loginform/LoginForm.js:
```diff
@@ -1,7 +1,37 @@
 
 function LoginForm() {
+  const dispatch = useDispatch();
+  const [email, setEmail] = useState('');
```
setEmail and setPassword is unused. And also email and password will be always "" . Why do you keep them here

becarselman commented on src/components/loginform/LoginForm.js:
```diff
@@ -1,7 +1,37 @@
+
+  const handleSubmit = (event) => {
+    event.preventDefault();
+    dispatch(loginRequest(email, password));
```
actions need to be called like actions.loginRequest(email, password)

to achieve this you need to have mapDispatchToProps. You can make separate file 
src/components/loginForm/modules/Container.js. and  connect mapStateToProps and mapDispatchToProps. Something like this.

![image](https://user-images.githubusercontent.com/102960105/232783890-c346fd44-5ed2-4b01-a6ce-97279223ad04.png)
and then you can connect that to your component with second file index.js
![image](https://user-images.githubusercontent.com/102960105/232784127-3c92a179-64af-4669-8ea7-ccaa01e3e6df.png)

becarselman commented on src/index.js:
```diff
@@ -1,17 +1,20 @@
-    <App />
-  </React.StrictMode>
+ReactDOM.render(
+  <Provider store={store}>
```
this need to be done in app.js not in index.js

becarselman commented on src/sagas/authSaga.js:
```diff
@@ -0,0 +1,19 @@
+  try {
+    const response = yield call(authService.login, action.payload.email, action.payload.password);
+    localStorage.setItem('token', response.data.token);
+    yield put({ type: LOGIN_SUCCESS });
```
you can import actions here and use like yield put(loginAction(data));

becarselman commented on src/service/authService.js:
```diff
@@ -0,0 +1,43 @@
+import { LOGIN_REQUEST, } from '../actions/types';
+
+
+function* loginSaga(action) {
```
what is this

becarselman commented on src/service/authService.js:
```diff
@@ -0,0 +1,43 @@
+function* loginSaga(action) {
+  try {
+    const { email, password } = action.payload;
+    const response = yield call(authService.login, email, password);
```
Why we use this function

becarselman replied:
and what is this

becarselman commented on src/service/authService.js:
```diff
@@ -0,0 +1,43 @@
+}
+
+
+const API_URL = 'http://localhost:3000/';
```
Please use env for this. env have purpose and we need to use it

becarselman commented on src/service/authService.js:
```diff
@@ -0,0 +1,43 @@
+const API_URL = 'http://localhost:3000/';
+
+const login = (email, password) => {
+  return axios.post(`${API_URL}/login`, { email, password })
```
use axios interseptors for requests. it will make you life easier

becarselman reviewed (changes requested):
This doesn't work at all. I tested and when I enter email and password nothing is happen. I know backend is not ready but you need to send request to backend.

TarikVelic committed:
FIXING THINGS FROM THE REVIEW

--------------------------------------------------------------------
 MAIN
 * Changed navigation
   Fixing things from the review
--------------------------------------------------------------------
 OTHER
 * ////
--------------------------------------------------------------------

becarselman commented on src/components/loginform/modules/Container.js:
```diff
@@ -0,0 +1,18 @@
+});
+
+const mapDispatchToProps = (dispatch) => ({
+  loginRequest: (email, password) => dispatch(loginRequest(email, password)),
```
Do this in one object. Because in this case if you have 30 actions you will need to pass 30 parameters in component props. But if you pass as object you will pass only 1. and then you can use actions.loginRequest()

becarselman commented on src/sagas/authSaga.js:
```diff
@@ -0,0 +1,38 @@
+} from "../actions/authActions";
+
+export function* loginSaga(action) {
+  yield put(loginRequest());
```
what is purpose of this line

becarselman commented on src/sagas/authSaga.js:
```diff
@@ -0,0 +1,38 @@
+  try {
+    const response = yield call(
+      authService.login,
+      action.payload.email,
```
same as prev. comment. Pass this parameters as object

becarselman commented on src/service/authService.js:
```diff
@@ -0,0 +1,58 @@
+function* loginSaga(action) {
+  try {
+    const { email, password } = action.payload;
+    const response = yield call(authService.login, email, password);
```
You resolved conversation and do nothing ???

becarselman commented on src/service/authService.js:
```diff
@@ -0,0 +1,58 @@
+}
+
+export function* watchAuth() {
+  yield takeLatest(actionTypes.LOGIN_REQUEST, loginSaga);
```
Why we need this

becarselman commented on src/service/authService.js:
```diff
@@ -0,0 +1,58 @@
+
+const API_URL = process.env.REACT_APP_API_URL;
+
+axios.interceptors.request.use(
```
Create intercepors in separate file and import here then

becarselman commented on src/service/authService.js:
```diff
@@ -0,0 +1,58 @@
+);
+
+const login = (email, password) => {
+  return axios.post(`${API_URL}/login`, { email, password })
```
you create interceptor and change nothing here ?

becarselman commented on src/service/authService.js:
```diff
@@ -0,0 +1,58 @@
+    });
+};
+
+const logout = () => {
```
you don't need logout as service function because you don't create any api call

becarselman commented on src/service/authService.js:
```diff
@@ -0,0 +1,58 @@
+};
+
+const isLoggedIn = () => {
+  return !!localStorage.getItem('userToken');
```
same

TarikVelic committed:
Review Changes

--------------------------------------------------------------------
 MAIN
 * Fixing stuff from review
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
Fixes from Review

--------------------------------------------------------------------
 MAIN
 * Fixing things and removing unnecessary stuff from the github review
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on src/App.js:
```diff
@@ -1,8 +1,22 @@
+
   return (
- <LoginForm></LoginForm>
+    <Router>
```
Move this into separate file  in src folder

becarselman commented on src/App.js:
```diff
@@ -1,8 +1,22 @@
+      <Routes>
+        <Route path="/" element={<Navigate to="/login" />} /> {/* Redirection from the home page to the login page */}
+        <Route path="/login" element={<LoginForm />} />
+        {/* Dodajte ostale rute ovdje */}
```
Please use english only

becarselman commented on src/actions/authActions.js:
```diff
@@ -0,0 +1,44 @@
+import { actionTypes } from "./types";
+
+export const loginUser = (email, password) => {
+  return {
```
use shorthand for function with only return

```
export const logoutUser = () => (

    type: actionTypes.LOGOUT_USER,
  
);
```

becarselman commented on src/actions/authActions.js:
```diff
@@ -0,0 +1,44 @@
+
+export const loginRequest = () => {
+  return {
+    type: actionTypes.AUTHENTICATE_USER_REQUEST,
```
Something missing here

becarselman commented on src/components/loginform/LoginForm.js:
```diff
@@ -1,37 +1,67 @@
+            type="email"
+            name="login_email"
+            value={email}
+            onChange={(e) => setEmail(e.target.value)}
```
move this function outside of jsx

becarselman commented on src/components/loginform/LoginForm.js:
```diff
@@ -1,37 +1,67 @@
+            type="password"
+            name="login_password"
+            value={password}
+            onChange={(e) => setPassword(e.target.value)}
```
same

becarselman commented on src/components/loginform/LoginForm.js:
```diff
@@ -1,37 +1,67 @@
+            className="placeholder"
+            placeholder="Enter your email"
+          />
+          <label className="passwordh2">Password</label>
```
change className into something better.  passwordh2 is bad option

becarselman commented on src/components/loginform/LoginForm.js:
```diff
@@ -1,37 +1,67 @@
+            name="login_email"
+            value={email}
+            onChange={(e) => setEmail(e.target.value)}
+            className="placeholder"
```
Placeholder is confusing for className

becarselman commented on src/sagas/authSaga.js:
```diff
@@ -0,0 +1,26 @@
+
+function* loginSaga(action) {
+  try {
+    const { email, password } = action.payload;
```
Here you destruct object And in second line you create same object.
Why you just don't use action.payload

becarselman commented on src/index.js:
```diff
@@ -1,17 +1,17 @@
-  <React.StrictMode>
+
+ReactDOM.render(
+  <Provider store={store}>
```
If you resolve conversation that mean you fix comment. In this case I don't see any fix. And also I need to mention this is not first time. Please change it because it is not a good habit. Sometimes you will get away without me noticing that you did it, but the goal of your internship is to teach you to work and not to learn to hide from me

becarselman commented on src/reducers/authReducer.js:
```diff
@@ -0,0 +1,26 @@
+        isAuthenticated: true,
+        user: action.payload
+      };
+    case actionTypes.LOGOUT_USER:
```
I will suggest better approach to this. Create actions setUser and use it for login, logout and all other places. On login you will set user with data and on logout in your case you will set as empty object

becarselman commented on src/sagas/authSaga.js:
```diff
@@ -0,0 +1,26 @@
+    const { email, password } = action.payload;
+    const response = yield call(authService.login, { email, password });
+    const { token, user } = response;
+    localStorage.setItem("jwtToken", token);
```
I think jwtToken is bad name for token. Use something more descriptive like accessToken.

becarselman commented on src/service/apiService.js:
```diff
@@ -0,0 +1,33 @@
+  (error) => {
+    if (error.response && error.response.status === 401) {
+      localStorage.removeItem('jwtToken');
+      window.location.reload();
```
Why you refresh page here?

becarselman commented on src/service/authService.js:
```diff
@@ -0,0 +1,24 @@
+    });
+};
+
+const logout = () => {
```
This is what I am talking in previous comment. You just resolve without fix

becarselman commented on src/service/authService.js:
```diff
@@ -0,0 +1,24 @@
+  localStorage.removeItem('jwtToken');
+};
+
+const isLoggedIn = () => {
```
Also why is this part of service.

becarselman commented on src/components/loginform/LoginForm.js:
```diff
@@ -1,37 +1,67 @@
+        navigate('/welcome');
+      })
+      .catch((error) => {
+        alert(error.message);
```
Please don't use alert. there are so many notification packages. Also with integration of some notification package you will reduce amount of work. On success login you will return notification success and store token, and for error you will return error notification. Automatically you don't need actions like loginSuccess and loginError

becarselman commented on src/components/loginform/LoginForm.js:
```diff
@@ -1,37 +1,67 @@
+  const handleSubmit = (event) => {
+    event.preventDefault();
+    actions.loginRequest(email, password)
+      .then(() => {
```
you can do this part in saga.

becarselman reviewed (changes requested):
Are you test this because I think it's not working

TarikVelic committed:
FIXING STUFF FROM REVIEW

--------------------------------------------------------------------
 MAIN
 * Translation of a couple of lines into English
   Moving unnecessary code from index to app.js

--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
FIXING REVIEW

--------------------------------------------------------------------
 MAIN
 * ///
--------------------------------------------------------------------
 OTHER
 * ///
--------------------------------------------------------------------

TarikVelic committed:
FIXES FROM REVIEW

--------------------------------------------------------------------
 MAIN
 * created a router.js file and moved the router from app.js there
   added payloud: null, in loginRequest
   deleted IsLoggedIn function
   I moved onChange={(e) => setEmail(e.target.value)} function outside of jsx
   and other...

--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
FEATURE/TariikMVP/DfWCIPVU

--------------------------------------------------------------------
 MAIN
 * Changed authReducer.js
--------------------------------------------------------------------

TarikVelic committed:
FEATURE/TariikMVP/DfWCIPVU

--------------------------------------------------------------------
 MAIN
 * Changed user into userId
--------------------------------------------------------------------

becarselman commented on src/components/loginform/LoginForm.js:
```diff
@@ -1,37 +1,81 @@
+    event.preventDefault();
+
+    try {
+      const res = await authService.login(email, password);
```
Why you are doing this here

becarselman commented on src/index.js:
```diff
@@ -1,17 +1,16 @@
-// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
-reportWebVitals();
+
+export default Container;
```
?

becarselman commented on src/index.js:
```diff
@@ -1,17 +1,16 @@
-import React from 'react';
-import ReactDOM from 'react-dom/client';
-import './index.css';
+import Container from './components/loginform/modules/Container';
```
unused

becarselman commented on src/index.js:
```diff
@@ -1,17 +1,16 @@
-import ReactDOM from 'react-dom/client';
-import './index.css';
+import Container from './components/loginform/modules/Container';
+import LoginForm from './components/loginform/LoginForm';
```
unused

becarselman commented on src/reducers/authReducer.js:
```diff
@@ -0,0 +1,42 @@
+  }
+}
+
+export const setUser = (user) => ({
```
?

becarselman commented on src/reducers/authReducer.js:
```diff
@@ -0,0 +1,42 @@
+  payload: user
+});
+
+export const login = (email, password) => async (dispatch) => {
```
?

becarselman commented on src/reducers/authReducer.js:
```diff
@@ -0,0 +1,42 @@
+  }
+};
+
+export const logout = () => (dispatch) => {
```
?

becarselman commented on src/reducers/errorReducer.js:
```diff
@@ -0,0 +1,17 @@
+  }
+}
+
+export const getErrors = (error) => ({
```
?

becarselman commented on src/sagas/authSaga.js:
```diff
@@ -0,0 +1,25 @@
+function* loginSaga(action) {
+  try {
+    const { email, password } = action.payloud;
+    const response = yield call(authService.login, email, password);
```
send email and password as 1 parameter( like object)

TarikVelic committed:
FEATURE/TariikMVP/DfWCIPVU
--------------------------------------------------------------------
 MAIN
 * Review changes
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
FEATURE/TariikMVP/DfWCIPVU

--------------------------------------------------------------------
 MAIN
 * Deleting unnecessary stuff from scss
   Deleting unnecessary codes from reducer

--------------------------------------------------------------------
 OTHER
 *
--------------------------------------------------------------------

TarikVelic committed:
FEATURE/TariikMVP/DfWCIPVU

--------------------------------------------------------------------
 MAIN
 * Updates in codes
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on src/components/loginform/LoginForm.js:
```diff
@@ -1,37 +1,89 @@
 import './LoginForm.scss';
 import Logotype from '../../images/loginform/Logotype.svg';
+import "../../index.css" 
```
why we have index.css import if we have /LoginForm.scss

becarselman commented on src/components/loginform/LoginForm.js:
```diff
@@ -1,37 +1,89 @@
+  };
+
+  const handleSubmit = (event) => {
+    console.log(actions.loginRequest);
```
remove console.log.

becarselman commented on src/components/loginform/LoginForm.js:
```diff
@@ -1,37 +1,89 @@
-      <div class="login-left">
-        <div class="image-container">
+    <div className="login-container">
+      {windowWidth >= 768 ? (
```
move ternary outside of jsx it will be more readable

TarikVelic committed:
FEATURE/TariikMVP/DfWCIPVU

--------------------------------------------------------------------
 MAIN
 * Small changes in code
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
FEATURE/TariikMVP/DfWCIPVU

--------------------------------------------------------------------
 MAIN
 * removing antd notifications
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

TarikVelic committed:
FEATURE/TariikMVP/DFWCIPVU

--------------------------------------------------------------------
 MAIN
 * Application JSON Header for CORS
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

becarselman commented on package.json:
```diff
@@ -6,9 +6,18 @@
+    "react-router": "^6.10.0",
+    "react-router-dom": "^6.10.0",
     "react-scripts": "5.0.1",
+    "react-toastify": "^9.1.2",
```
you have react-toastify and antd

becarselman commented on src/sagas/authSaga.js:
```diff
@@ -0,0 +1,27 @@
+function* loginSaga(action) {
+  try {
+    const { email, password } = action.payload;
+    const response = yield call(authService.login, { email, password });
```
you can send action.payload because it is {email, password}. You don't need to destruct in line 12 and send as object here

becarselman commented on src/sagas/authSaga.js:
```diff
@@ -0,0 +1,27 @@
+    localStorage.setItem("accessToken", token);
+    yield put(loginSuccess(userId)); 
+  } catch (error) {
+    console.log(error);
```
remove clg

becarselman commented on src/service/authService.js:
```diff
@@ -0,0 +1,14 @@
+    .then((response) => {
+      const { token, userId } = response.data;
+      localStorage.setItem('accessToken', token);
+      return userId;
```
where is .catch

TarikVelic committed:
FEATURE/TariikMVP/DfWCIPVU

--------------------------------------------------------------------
 MAIN
 * Removing unnecessary stuff from frontend
--------------------------------------------------------------------
 OTHER
 * <other msg>
--------------------------------------------------------------------

melcidar reviewed (approved):
Looks good to me :)
